adaptors_task/student/src/impl.cpp

```diff
@@ -1,8 +1,50 @@
 #include "impl.h"
 #include <vector>
+#include <stack>
+#include <algorithm>
 
 const std::vector<std::pair<char, char>> brackets {
     {'(', ')'},
     {'[', ']'},
     {'{', '}'}
 };
+
+bool isValid(const std::string& source)
+{
+    std::stack<char> stack;
+
+    for (auto& c : source)
+    {
+        auto it = std::find_if(brackets.begin(), brackets.end(), [&](const std::pair<char, char>& p) -> bool {
+            return p.first == c || p.second == c;
+        });
+
+        if (it == brackets.end())
+            continue; // Not a bracket
+
+        if (stack.empty())
+        {
+            stack.push(c);
+            continue;
+        }
+
+        if (it->first == c)
+        {
+            // Opening bracket
+            stack.push(c);
+        }
+        else if (it->second == c)
+        {
+            // Closing bracket
+            char lastChar = stack.top();
+
+            // Check if last opening bracket matches closing bracket
+            if (lastChar == it->first)
```

Comment:
Usually you should avoid comments in your code - everything should be shown in code, so please, just give a name to your conditions, so there will be no need for a comments